[JSC] Fix Compatibility Issues with `timezone` in `Intl.DateTimeFormat`

Gumichocopengin8 opened the pull request:
#### 1110ee18af6079baeb4144596c30d22b9d1c1625
<pre>
[JSC] Fix Compatibility Issues with `timezone` in `Intl.DateTimeFormat`
<a href="https://bugs.webkit.org/show_bug.cgi?id=295856">https://bugs.webkit.org/show_bug.cgi?id=295856</a>

Reviewed by Darin Adler.

The Intl.DateTimeFormat constructor [1] did not align with the current TC39 specification
regarding the handling of the timeZone option.
This patch implements GetAvailableNamedTimeZoneIdentifier [2] to bring the behavior
in line with the latest TC39 spec.

[1]: <a href="https://tc39.es/ecma402/#sec-createdatetimeformat">https://tc39.es/ecma402/#sec-createdatetimeformat</a>
[2]: <a href="https://tc39.es/ecma402/#sec-getavailablenamedtimezoneidentifier">https://tc39.es/ecma402/#sec-getavailablenamedtimezoneidentifier</a>

* JSTests/stress/intl-canonical-gmt.js:
* JSTests/stress/intl-datetimeformat.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::availableNamedTimeZoneIdentifier):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::canonicalizeTimeZoneName): Deleted.

Canonical link: <a href="https://commits.webkit.org/297594@main">https://commits.webkit.org/297594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fd28e3349f3ddf878bf95cd5ed222c7f13bcf73

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111996 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22152 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118019 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62233 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32343 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40244 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85079 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114943 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100777 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65511 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25145 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18918 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61869 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104484 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95214 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18994 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121336 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110555 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39029 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29036 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93925 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97036 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93742 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24012 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38948 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16741 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35051 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38923 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44435 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134815 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38560 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36297 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41888 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
```diff
@@ -739,15 +723,15 @@ void IntlDateTimeFormat::initializeDateTimeFormat(JSGlobalObject* globalObject,
     String timeZoneForICU;
     if (!tzValue.isUndefined()) {
-        String originalTz = tzValue.toWTFString(globalObject);
+        const String originalTz = tzValue.toWTFString(globalObject);
```
Why this change?

Gumichocopengin8 replied:
Because `originalTz` shouldn't be changed accidentally.

darinadler replied:
That’s true of thousands of other local variables in functions across WebKit, and as a matter of coding style we don’t use `const` for those. We could start but I don’t see why we’d start here.

Gumichocopengin8 replied:
Got it. I'll remove this `const`.

Gumichocopengin8 replied:
Updated

darinadler commented on Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
```diff
@@ -102,15 +102,13 @@ void IntlDateTimeFormat::setBoundFormat(VM& vm, JSBoundFunction* format)
 
-static String canonicalizeTimeZoneName(const String& timeZoneName)
+// https://tc39.es/ecma402/#sec-getavailablenamedtimezoneidentifier
+static String getAvailableNamedTimeZoneIdentifier(StringView timeZoneName)
```
WebKit coding style reserves “get” for only certain kinds of function names, so even though the standard calls this `GetAvailableNamedTimeZoneIdentifier` we should probably name this `availableNamedTimeZoneIdentifier`.

Gumichocopengin8 replied:
Fixed

darinadler commented on JSTests/stress/intl-datetimeformat.js:
```diff
@@ -254,31 +254,31 @@ shouldBe(Intl.DateTimeFormat('en', { timeZone: 'AMERICA/LOS_ANGELES' }).resolved
 shouldBe(Intl.DateTimeFormat('en', { timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone }).resolvedOptions().timeZone, Intl.DateTimeFormat().resolvedOptions().timeZone);
 
 // Time zone is canonicalized for obsolete links in IANA tz backward file.
```
This comment is no longer accurate. We should consider changing it. I’m also not sure we need all these tests if we’re just testing that the same string comes back to us. I wonder what this means about capitalization, like does “australia/darwin” work?

Gumichocopengin8 replied:
Updated the comment. 
I think these test cases are useful to make sure timezones are preserved and not canonicalized.

>  I wonder what this means about capitalization, like does “australia/darwin” work?

That's covered here

https://github.com/WebKit/WebKit/blob/d9db853f14ab390f3e69c476b831ae89b8e7f120/JSTests/stress/intl-datetimeformat.js#L249-L251

Gumichocopengin8 committed:
[JSC] Fix Compatibility Issues with `timezone` in `Intl.DateTimeFormat`
https://bugs.webkit.org/show_bug.cgi?id=295856

Reviewed by Darin Adler.

The Intl.DateTimeFormat constructor [1] did not align with the current TC39 specification
regarding the handling of the timeZone option.
This patch implements GetAvailableNamedTimeZoneIdentifier [2] to bring the behavior
in line with the latest TC39 spec.

[1]: https://tc39.es/ecma402/#sec-createdatetimeformat
[2]: https://tc39.es/ecma402/#sec-getavailablenamedtimezoneidentifier

* JSTests/stress/intl-canonical-gmt.js:
* JSTests/stress/intl-datetimeformat.js:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp:
(JSC::availableNamedTimeZoneIdentifier):
(JSC::IntlDateTimeFormat::initializeDateTimeFormat):
(JSC::canonicalizeTimeZoneName): Deleted.

Canonical link: https://commits.webkit.org/297594@main
